stock_picking_report_undelivered_product/tests/test_stock_picking_report_undelivered_product.py

```diff
@@ -0,0 +1,81 @@
+        self.ProductProduct = self.env['product.product']
+        self.StockPicking = self.env['stock.picking']
+        self.StockQuant = self.env['stock.quant']
+
+        self.warehouse = self.env.ref('stock.warehouse0')
+        self.stock_location = self.env.ref('stock.stock_location_stock')
+        self.customer_location = self.env.ref('stock.stock_location_customers')
+        self.picking_type_out = self.env.ref('stock.picking_type_out')
+
+        self.partner_display = self.ResPartner.create({
+            'name': 'Partner for test display',
+            'customer': True,
+            'display_undelivered_in_picking': True,
+        })
+        self.partner_no_display = self.ResPartner.create({
+            'name': 'Partner for test on display',
+            'customer': True,
+            'display_undelivered_in_picking': False,
+        })
+
+        self.product_display = self.ProductProduct.create({
+            'name': 'Test undelivered display',
+            'display_undelivered_in_picking': True,
+            'type': 'product',
+        })
+        self.product_no_display = self.ProductProduct.create({
+            'name': 'Test undelivered no display',
+            'display_undelivered_in_picking': False,
+            'type': 'product',
+        })
+        self.StockQuant.create({
+            'product_id': self.product_no_display.id,
+            'location_id': self.warehouse.lot_stock_id.id,
+            'quantity': 2000,
+        })
+
+    def _create_picking(self, partner):
+        picking_form = Form(self.StockPicking)
+        picking_form.picking_type_id = self.picking_type_out
+        picking_form.partner_id = partner
+
+        with picking_form.move_ids_without_package.new() as line:
+            line.product_id = self.product_display
+            line.product_uom_qty = 50.00
+        with picking_form.move_ids_without_package.new() as line:
+            line.product_id = self.product_no_display
+            line.product_uom_qty = 20.00
+        return picking_form.save()
+
+    def test_displayed_customer(self):
+        picking = self._create_picking(self.partner_display)
+        picking.action_confirm()
+        picking.action_assign()
+        picking.move_line_ids.qty_done = 10.00
+        picking.action_done()
+        res = self.env['ir.actions.report']._get_report_from_name(
+            'stock.report_deliveryslip'
+        ).render_qweb_html(picking.ids)
+        self.assertRegexpMatches(str(res[0]), "undelivered_product")
+
+    def test_no_displayed_customer(self):
+        picking = self._create_picking(self.partner_no_display)
+        picking.action_confirm()
+        picking.action_assign()
+        picking.move_line_ids.qty_done = 10.00
+        picking.action_done()
+        res = self.env['ir.actions.report']._get_report_from_name(
+            'stock.report_deliveryslip'
+        ).render_qweb_html(picking.ids)
+        self.assertNotRegex(str(res[0]), "undelivered_product")
```

Comment:
Not sure about the purpose of this assert, but if it's that the text is not found, then please use:

```suggestion
        self.assertNotIn("undelivered_product", str(res[0]))
```